Merge latest RUC LSM into community develop

tanyasmirnova opened the pull request:
1. Updated computation of snow cover fraction and snow thermal conductivity
2. Other changes that are implemented in the real-time RRFS_B.

tanyasmirnova committed:
Merged latestest version of RUC LSM into community develop.

tanyasmirnova committed:
Update parameters for RUC LSM.

tanyasmirnova committed:
Switch to the default options in snow model

tanyasmirnova committed:
Moved some variables from Interstitial to Sfcprop

dustinswales commented on physics/namelist_soilveg_ruc.F90:
Is there a need for the namelist declarations to be in a separate module from the routine reading the namelist? 
It would be best if these were moved to set_soilveg_ruc.F90. The[ gfdl microphysics](https://github.com/ufs-community/ccpp-physics/blob/ufs/dev/physics/module_gfdl_cloud_microphys.F90), and other schemes, keep these pieces together.

tanyasmirnova replied:
@dustinswales I agree, but I followed the Noah LSM configuration, maybe this could be changed later.

dustinswales commented on physics/module_sf_ruclsm.F90:
```diff
@@ -19,28 +19,53 @@ MODULE module_sf_ruclsm
 
 !> CONSTANT PARAMETERS
+!! @{
       real (kind=kind_phys), parameter :: P1000mb = 100000.
```
Most of these are already defined in [physcons.F90](https://github.com/ufs-community/ccpp-physics/blob/ufs/dev/physics/physcons.F90). It's ideal to pass these into the schemes directly, or at least pass the constants to the _init() phase and set internal module variables that are used throughout the scheme.

tanyasmirnova replied:
The change is done to use constants from physcons.F90.

dustinswales commented on physics/lsm_ruc.F90:
```diff
@@ -995,14 +1010,14 @@ subroutine lsm_ruc_run                                            & ! inputs
           soilt1_lnd(i,j) = tsurf_lnd(i)
         endif
-        tsnav_lnd(i,j) = 0.5*(soilt_lnd(i,j) + soilt1_lnd(i,j)) - 273.15
+        tsnav_lnd(i,j) = min(0.,0.5*(soilt_lnd(i,j) + soilt1_lnd(i,j)) - 273.15)
```
Replace **min(0.** with **min(0._kind_phys**

tanyasmirnova replied:
@dustinswales and also 0.5 to 0.5_kind_phys?

dustinswales replied:
It's best if all constants had the _kind_phys suffix. This way there is no ambiguity about the precision, making the code more portable. It's possible to lose precision w/o notification from the compiler, so it's best to be explicit. https://www.fortran90.org/src/best-practices.html

tanyasmirnova replied:
This line is changed. There are many constants in the code. I haven't added kind_phys to all of them yet.

dustinswales commented on physics/lsm_ruc.F90:
```diff
@@ -1018,27 +1033,79 @@ subroutine lsm_ruc_run                                            & ! inputs
-          if(sneqv_lnd(i,j)/snowh_lnd(i,j) > 950.) then
-            sneqv_lnd(i,j) = 300. * snowh_lnd(i,j)
+        !> -- sanity checks on sneqv and snowh
+        if (sneqv_lnd(i,j) /= 0.0d0 .and. snowh_lnd(i,j) == 0.0d0) then
```
If these need (?) to be doubles, I would suggest importing kind_dbl_prec, alongside kind_phys, from machine.F90.

tanyasmirnova replied:
kind_dbl_prec is used here instead of d0.

dustinswales commented on physics/lsm_ruc.F90:
```diff
@@ -787,6 +790,19 @@ subroutine lsm_ruc_run                                            & ! inputs
         frpcpn = .false.
       endif
 
+      do j  = 1, 1    ! 1:1
```
It's probably a good idea to parameterize this "1", used here and below for the "j" loop extent. Also in the declaration of fields that have a "j" dimension.
For now it could be defined in this module, but eventually could be provided as an input from the host.

tanyasmirnova replied:
1,1 is changed to jms,jme.

dustinswales commented on physics/module_sf_ruclsm.F90:
```diff
@@ -879,36 +911,48 @@ SUBROUTINE LSMRUC(                                           &
 
-        smtotold(i,j)=smtotold(i,j)+(qmin+soilm1d(nzs))*           &
-                    (zsmain(nzs)-zshalf(nzs))
+        !smtotold(i,j)=smtotold(i,j)+(qmin+soilm1d(nzs))*           &
```
No need to comment code with github history.

tanyasmirnova replied:
Sometimes I keep commented out lines for my memory, but you are right, it does make the code look messy. Removed the commented lines.

dustinswales commented on physics/module_sf_ruclsm.F90:
```diff
@@ -984,25 +1026,38 @@ SUBROUTINE LSMRUC(                                           &
-                    (zsmain(nzs)-zshalf(nzs))
-        smmax (i,j) =smmax (i,j)+(qmin+dqm)*                     &
-                    (zsmain(nzs)-zshalf(nzs))
+        !smavail(i,j)=smavail(i,j)+(qmin+soilm1d(nzs))*           &
```
ditto

tanyasmirnova replied:
removed

dustinswales commented on physics/module_sf_ruclsm.F90:
```diff
@@ -1022,7 +1077,7 @@ SUBROUTINE LSMRUC(                                           &
 !30july13        qcg(i,j)=qcg(i,j)+dew(i,j)/qkms
 
-        Z0       (I,J) = ZNT (I,J)
+        !Z0       (I,J) = ZNT (I,J)
```
ditto

tanyasmirnova replied:
uncommented

dustinswales commented on physics/module_sf_ruclsm.F90:
```diff
@@ -1074,9 +1126,9 @@ SUBROUTINE LSMRUC(                                           &
-       if(snowfrac > 0. .and. xice(i,j).ge.xice_threshold ) then
-           SNOWFRAC = SNOWFRAC*XICE(I,J)
-       endif
+       !if(snowfrac > 0. .and. xice(i,j).ge.xice_threshold ) then
```
ditto

tanyasmirnova replied:
this commented line is removed.

dustinswales commented on physics/module_sf_ruclsm.F90:
```diff
@@ -1121,27 +1187,29 @@ SUBROUTINE LSMRUC(                                           &
-  print *,'CANWATold, canwat ',i,j,canwatold(i,j),canwat(i,j)
-  print *,'canwat(i,j)-canwatold(i,j)',max(0.,canwat(i,j)-canwatold(i,j))
-    ENDIF
+!  print *,'Smf=',smf(i,j),i,j
```
ditto

tanyasmirnova replied:
removed

dustinswales reviewed (dismissed):
I know this PR is still a draft, but we thought it was time it got some attention.
I have some suggestions and if you like, I can open a PR into this branch in the coming days?

tanyasmirnova committed:
Taking care of Dustin Wales comments, removing unnecessary comments,
cleaning, etc.

tanyasmirnova committed:
Address reviewers comments in the RUC LSM driver.

tanyasmirnova committed:
Clean-up print statements.

tanyasmirnova committed:
Added table parameters for option 3 of snow cover fraction computation.

tanyasmirnova committed:
Updated the sfc_diag.f that computes 2-m diagnostics. Should not
affect results for physics suites not using RUC LSM.

tanyasmirnova committed:
In module_sf_ruclsm.F90 changes rellated to kind_physics.

tanyasmirnova committed:
Fixed the problem in sfc_daig.f: it was using
"surface_exchange_coefficient_for_heat_at_2m" which is provided only by
MYNN surface layer scheme and not the others. Now this variable is
comuted internally in sfc_diag.f.

tanyasmirnova committed:
Continue changes for kind_phys with constants.

tanyasmirnova committed:
More changes related to kin_phys.

tanyasmirnova committed:
More changes in RUC LSM related to kind_phys.

grantfirl commented:
@tanyasmirnova I noticed that this PR is undoing some recent changes in ccpp-physics. Can you please merge the latest ufs/dev branch into your PR branch to make reviewing easier?

grantfirl commented on physics/GFS_debug.F90:
```diff
@@ -1327,7 +1332,7 @@ subroutine GFS_interstitialtoscreen_run (Model, Statein, Stateout, Sfcprop, Coup
                      call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Interstitial%tprcp_land          ', Interstitial%tprcp_land              )
                      call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Interstitial%tprcp_water         ', Interstitial%tprcp_water             )
-                     call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Interstitial%trans               ', Interstitial%trans                   )
+                     call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Diag%trans               ', Interstitial%trans                   )
```
If this variable is now in the Diag DDT, its `print_var` should get moved to GFS_diagtoscreen_run.

tanyasmirnova replied:
done

tanyasmirnova replied:
This variable is moved to GFS_diagtoscreen_run.

grantfirl commented on physics/GFS_debug.F90:
```diff
@@ -1315,6 +1317,9 @@ subroutine GFS_interstitialtoscreen_run (Model, Statein, Stateout, Sfcprop, Coup
                      call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Interstitial%sigmatot            ', Interstitial%sigmatot                )
                      call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Interstitial%snowc               ', Interstitial%snowc                   )
+                     call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Interstitial%snowd_ice           ', Interstitial%snowd_ice               )
+!                    call print_var(mpirank, omprank, blkno, Grid%xlat_d, Grid%xlon_d, 'Interstitial%snowd_land          ', Interstitial%snowd_land              )
```
If the variables associated with the commented-out `print_var` statements in this file have been moved from interstitial to other DDTs permanently, please just delete the lines instead of commenting them out.

tanyasmirnova replied:
These variables are not moved from interstitial. Uncommented snowd_land and removed snowd_water (which is always zero).

tanyasmirnova replied:
One line here is uncommented, and the second one is removed.

grantfirl commented on physics/lsm_ruc.F90:
```diff
@@ -365,18 +367,20 @@ subroutine lsm_ruc_run                                            & ! inputs
-      real(kind=kind_phys), parameter :: rhoh2o  = 1000.0
-      real(kind=kind_phys), parameter :: stbolt  = 5.670400e-8
+      real(kind_phys), parameter :: rhoh2o  = 1000.0
+      real(kind_phys), parameter :: stbolt  = 5.670400e-8
```
This should be passed in from the host:
`[stbolt]`
 `standard_name = stefan_boltzmann_constant`
 `long_name = Stefan-Boltzmann constant`
 `units = W m-2 K-4`
 `dimensions = ()`
 `type = real`
 `kind = kind_phys`
 `intent = in`

tanyasmirnova replied:
Changed to pass it from the host.

tanyasmirnova replied:
Use con_sbc for Stefan-Boltzman constant.

grantfirl commented on physics/lsm_ruc.F90:
```diff
@@ -365,18 +367,20 @@ subroutine lsm_ruc_run                                            & ! inputs
 !  ---  constant parameters:
-      real(kind=kind_phys), parameter :: rhoh2o  = 1000.0
-      real(kind=kind_phys), parameter :: stbolt  = 5.670400e-8
+      real(kind_phys), parameter :: rhoh2o  = 1000.0
```
This should be passed in from the host:
`[rhoh20]`
 `standard_name = fresh_liquid_water_density_at_0c`
 `long_name = density of liquid water`
 `units = kg m-3`
 `dimensions = ()`
 `type = real`
 `kind = kind_phys`
 `intent = in`

tanyasmirnova replied:
A change is done to use the water density from the host.

grantfirl commented on physics/lsm_ruc.F90:
```diff
@@ -1060,81 +1115,50 @@ subroutine lsm_ruc_run                                            & ! inputs
-            write (0,*)'shdmax1d(i,j) =',i,j,shdmax1d(i,j)
-            write (0,*)'rdlai2d =',rdlai2d
-          endif
+        !if (debug_print) then
```
Why was `if (debug_print) then` commented out? I doubt everyone wants to see this output for the first 10 timesteps every time they run, right?

tanyasmirnova replied:
Comment is removed.

grantfirl commented on physics/lsm_ruc.meta:
```diff
@@ -1344,13 +1398,37 @@
 [snowfallac_lnd]
-  standard_name = surface_snow_amount_over_land
-  long_name = run-total snow accumulation on the ground over land
+  standard_name = surface_snow_amount_vardens_over_land
```
I appreciate adding specificity to the metadata, but we try not to have arbitrary contractions (e.g. `varsdens`) in standard names. Could this (and also the similar ice-equivalent) be changed to `surface_snow_amount_assuming_variable_snow_density_over_land`?

tanyasmirnova replied:
The change to surface_snow_amount_assuming_variable_snow_density_over_land is done.

grantfirl commented on physics/lsm_ruc.meta:
```diff
@@ -1527,6 +1621,7 @@
   type = real
   kind = kind_phys
   intent = inout
+  active = (control_for_land_surface_scheme == identifier_for_ruc_land_surface_scheme)
```
The active attribute should only be added to host-side metadata. Please remove this.

tanyasmirnova replied:
Removed.

grantfirl commented on physics/sfc_diag.f:
```diff
@@ -1,39 +1,62 @@
 !>  \file sfc_diag.f
```
I'm not sure what is going on with the changes in this file, but it looks like it's missing changes from previous commits. See https://github.com/ufs-community/ccpp-physics/commit/6f1c2b1188737721b30bf4a98e432792a1c24742#diff-5210406169e53c367678af29f850545c25156baf7a3f70ab811ea9adba8b7cae. I wonder if you could start from the latest version in ufs/dev and put your changes back in?

tanyasmirnova replied:
There was some confusion with this code. It is ok now.

grantfirl commented on physics/sfc_diag.meta:
```diff
@@ -157,6 +226,44 @@
   type = logical
   intent = in
+[diag_flux]
+  standard_name = flag_for_flux_method_in_2m_diagnostics
```
Could you change 'flag_for_X' to `do_X` for this new variable's standard name and the one below?

tanyasmirnova replied:
I followed here the common practice in GFS_typedefs.F90 to call such control variables as flags.

grantfirl commented on physics/module_sf_ruclsm.F90:
```diff
@@ -19,28 +19,56 @@ MODULE module_sf_ruclsm
-      real (kind=kind_phys), parameter :: r_v     = 4.6150e+2
+!! @{
+      real (kind_phys), parameter :: P1000mb = 100000._kind_dbl_prec
+      real (kind_phys), parameter :: xls     = 2.85E6_kind_dbl_prec
```
To be consistent with the host model and other physics, these constants should not be redefined. `xls` should be the sum of `latent_heat_of_vaporization_of_water_at_0C` and `latent_heat_of_fusion_of_water_at_0C`.

tanyasmirnova replied:
Done.

grantfirl commented on physics/module_sf_ruclsm.F90:
```diff
@@ -19,28 +19,56 @@ MODULE module_sf_ruclsm
+!! @{
+      real (kind_phys), parameter :: P1000mb = 100000._kind_dbl_prec
+      real (kind_phys), parameter :: xls     = 2.85E6_kind_dbl_prec
+      real (kind_phys), parameter :: rhowater= 1000._kind_dbl_prec
```
This is defined as `fresh_liquid_water_density_at_0c`

tanyasmirnova replied:
Now the code uses constants from Physcons.

grantfirl commented on physics/module_sf_ruclsm.F90:
```diff
@@ -19,28 +19,56 @@ MODULE module_sf_ruclsm
+      real (kind_phys), parameter :: P1000mb = 100000._kind_dbl_prec
+      real (kind_phys), parameter :: xls     = 2.85E6_kind_dbl_prec
+      real (kind_phys), parameter :: rhowater= 1000._kind_dbl_prec
+      real (kind_phys), parameter :: piconst = 3.1415926535897931_kind_dbl_prec
```
This is defined as `pi`.

tanyasmirnova replied:
Now the code uses constants from Physcons.

grantfirl commented on physics/module_sf_ruclsm.F90:
```diff
@@ -19,28 +19,56 @@ MODULE module_sf_ruclsm
+      real (kind_phys), parameter :: xls     = 2.85E6_kind_dbl_prec
+      real (kind_phys), parameter :: rhowater= 1000._kind_dbl_prec
+      real (kind_phys), parameter :: piconst = 3.1415926535897931_kind_dbl_prec
+      real (kind_phys), parameter :: r_v     = 461.50_kind_dbl_prec
```
This is defined as `gas_constant_water_vapor`

tanyasmirnova replied:
Now the code uses constants from Physcons.

grantfirl commented on physics/module_sf_ruclsm.F90:
```diff
@@ -19,28 +19,56 @@ MODULE module_sf_ruclsm
+      real (kind_phys), parameter :: r_v     = 461.50_kind_dbl_prec
+      real (kind_phys), parameter :: zero    = 0._kind_dbl_prec
+      real (kind_phys), parameter :: one     = 1._kind_dbl_prec
+      real (kind_phys), parameter :: tfrz    = 273.15_kind_dbl_prec
```
This is defined as `temperature_at_zero_celsius`

tanyasmirnova replied:
Now the code uses constants from Physcons.

grantfirl commented on physics/module_sf_ruclsm.F90:
```diff
@@ -19,28 +19,56 @@ MODULE module_sf_ruclsm
-      real (kind=kind_phys), parameter :: piconst = 3.1415926535897931
-      real (kind=kind_phys), parameter :: r_v     = 4.6150e+2
+!! @{
+      real (kind_phys), parameter :: P1000mb = 100000._kind_dbl_prec
```
There is a clash between using using `kind_phys` and `kind_dbl_prec`. I'd stick with `kind_phys` like the rest of physics.

tanyasmirnova replied:
Some subroutines use double precision constants. For example, sfc_diag_post.f

tanyasmirnova committed:
In GFS_diag.F90 moved Trans variable from Diag DDT to GFS_diagtoscreen. Uncommented snowd_land, and
removed snowd_water.

grantfirl commented on physics/module_sf_ruclsm.F90:
Thanks for cleaning up a lot of the commented-out lines in this file!

tanyasmirnova replied:
I like my code much better now. Thank you for a careful review.

tanyasmirnova committed:
More changes related to kind_phys. Use constants from Physcons.

tanyasmirnova committed:
Use constants from host in RUC LSM driver.

tanyasmirnova committed:
Removed wet - not used. Also rename rocp into con_rocp

tanyasmirnova committed:
More changes in RUC LSM related to kind_phys and use of constants from
Physcons.

tanyasmirnova committed:
Changes related to kind_phys.

tanyasmirnova committed:
Moved RUC LSM options to namelist parameters. Also, some clean-up of
print statements.

tanyasmirnova committed:
Surface_snow_amount_vardens_over_land is changed to surface_snow_amount_vardens_over_land
Same for ice.

tanyasmirnova committed:
Active = (control_for_land_surface_scheme == identifier_for_ruc_land_surface_scheme) is removed.

SamuelTrahanNOAA committed:
Merge remote-tracking branch 'upstream/ufs/dev' into ufsdev_ruclsm

joeolson42 committed:
Updates to the MYNN surface-layer scheme

tanyasmirnova committed:
Modifications to 2-m diagnostics will affect the results only with the use of RUC LSM.

tanyasmirnova committed:
Use q1(i) instead of q1c in original diagnostics.

tanyasmirnova committed:
Replaced double precision zero with 0. as it was in the original version.

joeolson42 committed:
Merge branch 'ufs/dev' into mynnsfc_24March

SamuelTrahanNOAA committed:
Merge remote-tracking branch 'upstream/ufs/dev' into ufsdev_ruclsm

tanyasmirnova committed:
Removed mosaic_lu and mosaic_soil from namelist_soilveg_ruc and set_soilveg_ruc as they are moved to the namelist options in GFS_typedefs.F90.

SamuelTrahanNOAA committed:
Merge remote-tracking branch 'joe/mynnsfc_24March' into ufsdev_ruclsm

zach1221 commented:
Hello, I think we're ready to continue working towards merging [#1646](https://github.com/ufs-community/ufs-weather-model/pull/1646). @tanyasmirnova are you able to resolve the conversations in this PR and request final approval?

tanyasmirnova commented:
@zach1221 I resolved all reviewer's comments and would like to get final approvals from reviewers.

grantfirl reviewed (approved):
We'll clean up use of physcons later since all testing has been done.

jkbk2004 commented:
@grantfirl all tests are done: https://github.com/ufs-community/ufs-weather-model/pull/1646. can you go ahead to merge this pr? @dustinswales can you approve the pr?
